.github/workflows/publish.yml

```diff
@@ -0,0 +1,132 @@
+name: publish
+
+on:
+  push:
+    tags: '\d+.\d+.\d+'
+
+env:
+  POETRY_VERSION: 1.6
+
+jobs:
+  on-main-branch-check:
+    runs-on: ubuntu-latest
+    outputs:
+      on_main: ${{ steps.contains_tag.outputs.retval }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - uses: rickstaa/action-contains-tag@v1
+        id: contains_tag
+        with:
+          reference: "main"
+          tag: "${{ github.ref_name }}"
+
+  tests:
+    name: tests
+    needs: on-main-branch-check
+    if: ${{ needs.on-main-branch-check.outputs.on_main == 'true' }}
+    uses: "./.github/workflows/tests.yml"
+
+  build-wheels:
+    name: build wheels
+    needs: tests
+    uses: "./.github/workflows/wheels.yml"
+
+  build-sdist:
+    name: build source distribution
+    needs: tests
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+          submodules: true
+
+      - uses: actions/setup-python@v5
+        with:
+          python-version: 3.12
```

Comment:
Apologies for the ignorant question, but in `tests.yml` we support `["3.8", "3.9", "3.10", "3.11"]` but not 3.12.  

1.  Should we add `3.12` to the list of versions to test against?
2.  Do we need to build an `sdist` _for each_ python version?

Reply:
1. Yes! I'll add Py3.12 and if tests pass, then we should keep it.
2. No, we only need one `sdist` so it doesn't matter which Python we use to build it.